pages/docs/guides/message-validation.md

```diff
@@ -0,0 +1,118 @@
+
+An example implementation of message validation in a gateway is the [AsyncAPI gateway](https://github.com/asyncapi/event-gateway). It intercepts all incoming messages moving them into a pipeline of middlewares and handlers such as message validation. You can use a Kafka consumer/producer[(kcat)](https://github.com/edenhill/kcat), a broker, and a simple WebSocket to run the AsyncAPI gateway in your machine.
+Check out an [AsyncAPI file demo with Studio.](https://studio.asyncapi.com/?url=https://raw.githubusercontent.com/asyncapi/event-gateway/master/deployments/k8s/event-gateway-demo/event-gateway-demo.asyncapi.yaml)
+
+<Remember>
+Currently, only the Kafka protocol is supported.
+</Remember>
+
+```mermaid
+graph TD
+    PR[Producer]-- Message --- EG[AsyncAPI Event-Gateway]
+    EG -- Message --- EGV{Is Message valid?}
+    EGV -->|Yes| BR[Broker]
+    EGV --- |No| INV{Fail when invalid?}
+    INV -->|Yes| ERR[/Fail/] -- Produce request errored --> PR
+    INV -->|No| BR
+```
+The AsyncAPI document is important in this case because payload schemas are taken from it and messages as validated against it in your application.
+You can spin up the AsyncAPI gateway using an AsyncAPI file. All the messages are forwarded to a WebSocket endpoint; if the message/payload is invalid, it includes a validation error message.
+
+### Validation handled by the broker
+As producers and consumers do not communicate with each other directly, but rather information transfer happens via Kafka topic. At the same time, the consumer still needs to know the type of data the producer is sending. Imagine if the producer starts sending bad data to Kafka or if the data type of your data gets changed. We need a way to have a common data type that must be agreed upon.
+
+That’s where Schema Registry comes into the picture. It is an application that resides outside of your Kafka cluster and handles the distribution of schemas to the producer and consumer by storing a copy of schema in its local cache.
+
+```mermaid
+graph LR
+    A([Producer]) --->|send schemaID| B(Schema Registry) & C{Kafka Cluster}--->|recieve schemaID| D([Consumer])
+```
+With the schema registry in place, the producer first talks to the schema registry and checks if the schema is available before sending the data to Kafka. If it cannot locate the schema, it registers and caches it in the schema registry. When the producer receives the schema,to sends it to Kafka prefixed with a unique schema ID. When the consumer processes this message, it will communicate with the schema registry using the schema ID obtained from the producer. If there is a schema mismatch, the schema registry will throw an error, informing the producer that it is violating the schema agreement.
+
+```mermaid
+stateDiagram-v2
```

Comment:
it's giving error when I did that

Reply:
@starlightknown Please add a comment where showing @derberg the error message. 😄 It is not helpful to him if you only tell him you get an error; he (we) needs to see the error to help you troubleshoot.

Reply:
![image](https://user-images.githubusercontent.com/74637789/207508201-71046e12-3cb6-4b51-9899-9fc4123f94d9.png)

**original**

![image](https://user-images.githubusercontent.com/74637789/207508388-5007a053-9f29-4a04-b1ef-ab7555689b35.png)

**with graph TD**

Reply:
better share a link so I can try in mermaid.

anyway, errors happen as the syntax of `stateDiagram` is not the same as `graph`, so it is not enough to just change the name

Reply:
look what I got from https://chat.openai.com/chat

my question was: `generate for me mermaid.js graph diagram that explains how schema validation works in kafka with schema registry`

answer is mindblowing:
> I would only change `Retrieve message` to `Receive message` and explanation that consumer receives and not retrieves message

Here is an example of a mermaid.js graph diagram that explains how schema validation works in Kafka with the Schema Registry:

```mermaid
sequenceDiagram
Producer ->> Schema Registry: Validate message schema
Schema Registry ->> Producer: Return validation result

Producer ->> Kafka: Publish message
Kafka ->> Consumer: Retrieve message
Consumer ->> Schema Registry: Validate message schema
Schema Registry ->> Consumer: Return validation result
```

Explanation:

1. The producer sends the message schema to the Schema Registry to be validated.
1. The Schema Registry returns the validation result to the producer.
1. If the schema is valid, the producer publishes the message to Kafka.
1. The consumer retrieves the message from Kafka.
1. The consumer sends the message schema to the Schema Registry to be validated.
1. The Schema Registry returns the validation result to the consumer.

This diagram shows how the Schema Registry is used to validate the schemas of messages being produced and consumed in Kafka. The producer sends the message schema to the Schema Registry for validation before publishing the message to Kafka, and the consumer does the same before processing the message. This helps ensure that the messages being exchanged between producers and consumers conform to the agreed-upon schema, which can help prevent errors and improve data integrity.



![Screenshot 2022-12-21 at 11 43 32](https://user-images.githubusercontent.com/6995927/208886712-692f98d6-b6ba-4045-aa94-48746a45c473.png)

Reply:
after another run it said `Consume message` instead of `Retrieve message`

Reply:
Very interesting, I think retrieve fits more? Because consumer consumes message is a bit general ig

Reply:
also, I was wrong, I always thought that is it broker that communicates with schema reqistry, but it is producers and consumer responsibility, so we should rename section from `Validation handled by the broker` to `Validation using Schema Registry`

Reply:
alternative answer from openAPI that is much shorter:

```mermaid
graph LR
A[Producer] --> B(Kafka Broker)
B --> C[Consumer]
C --> D{Does the schema match?}
D -- Yes --> E[Process message]
D -- No --> F[Discard message]
```

Here's what each component does:

* Producer: This is the component that produces messages and sends them to the Kafka broker.
* Kafka Broker: This is the component that receives messages from the producer and stores them in Kafka topics.
* Consumer: This is the component that reads messages from Kafka topics and processes them.
* Schema Registry: This is a component that stores schemas for messages and enables schema validation.

When a producer sends a message to the Kafka broker, the broker stores the message in the appropriate topic. When a consumer reads the message, it checks the message's schema against the schema stored in the schema registry. If the schema matches, the consumer processes the message. If the schema does not match, the consumer discards the message.

Reply:
> Very interesting, I think retrieve fits more? Because consumer consumes message is a bit general ig

Retrieve suggests (at least for me non native speaker) that consumer communicates with broker to read message, but in fact it is broker that sends a message, so consumer receives/consumes it

Reply:
> also, I was wrong, I always thought that is it broker that communicates with schema reqistry, but it is producers and consumer responsibility, so we should rename section from `Validation handled by the broker` to `Validation using Schema Registry`

ah actually we did this heading before and changed it again XD I'll fix it

Reply:
so shall we go with the longer version of the diagram or the shorter one from openapi? I feel like the first one shows more in detail about what's happening in between the schema registry, producer, consumer. The second one is similar to the diagram which is already present in the guide now.